### PR TITLE
core: using native interrupt handlers

### PR DESCRIPTION
**Pull request by etienne-lms**
Adds a section in core architecture documentation to describe interrupt controllers API functions.
These commits are based on stalled P-R https://github.com/OP-TEE/optee_docs/pull/192.

The 1st commit describes the API since https://github.com/OP-TEE/optee_os/pull/6354 has been merged.
The 2nd commit is an RFC describing the API as per P-R https://github.com/OP-TEE/optee_os/pull/6361.

### CODE REVIEW

File: architecture/core.rst [hunk 1]
```diff
@@ -390,6 +390,140 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+
+An interrupt controller driver initializes its ``struct itr_chip`` instance
+by calling API function ``itr_chip_init()``.
+
+A reference to CPU main interrupt controller, e.g. a GIC instance on Arm
+architecture CPUs, can be retrieved with API function
+``interrupt_get_main_chip()``. Main interrupt controller driver must
+register as main controller using API function ``interrupt_main_init()``
+which is in charge of calling ``itr_chip_init()``.
+
```
Comment (etienne-lms) [inline architecture/core.rst:468]: Info to add:
> Interrupt controller shall call `interrupt_call_handlers(chip, num)` in order to call the registered interrupt consumer(s) handler(s).

**Review by etienne-lms — COMMENTED** [review]
@jbech-linaro, i think the 1st change should be merged before next tag. It descrides the existing implmentation. I'll remove the 2nd commit and will post it upon https://github.com/OP-TEE/optee_os/pull/6361 completion.

**Comment by etienne-lms** [comment]
I have remove the description related to P-R #6361, used a specific section for main controller info, rephrased a bit the description and add `interrupt_call_handlers()` info.
I think this change should in next tag documentation.

File: architecture/core.rst [hunk 2]
```diff
@@ -390,6 +390,145 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+An interrupt chip driver registers the controller instance with API
+function ``itr_chip_init()``. The driver calls the registered interrupt
+consumer(s) handler(s) with API function ``interrupt_call_handlers()``.
+
+**CPU main interrupt controller driver**
+
+CPU main interrupt controller (e.g. a GIC instance on Arm architecture CPUs)
+driver must register itself as main controller using API function
+``interrupt_main_init()``. The functiob is in charge of calling
```
Comment (jenswi-linaro) [inline architecture/core.rst:446]: s/functiob/function/

File: architecture/core.rst [hunk 3]
```diff
@@ -390,6 +390,145 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+``itr_chip_init()`` for that chip instance.
+
+Interrupt consumer drivers get can get the CPU main interrupt controller
+reference with API function ``interrupt_get_main_chip()``.
+
+**Interrupt handlers**
+
+Interrupt handler functions are functions registered by interrupt consumer
```
Comment (jenswi-linaro) [inline architecture/core.rst:454]: "are callback functions" perhaps?

File: architecture/core.rst [hunk 4]
```diff
@@ -390,6 +390,145 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+**Interrupt management API functions**
+
+Interrupt management resources are declared in header file
+kernel/interrupt.h_. Interrupt consumers main API functions are:
+
+    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
+      enable and disable an interrupt.
+
+    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
+      mask/unmask an interrupt.
```
Comment (jenswi-linaro) [inline architecture/core.rst:418]: mask or unmask

I think we need to mention the difference between mask/unmask and enable/disable.

File: architecture/core.rst [hunk 5]
```diff
@@ -390,6 +390,145 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+by that controller. In the literature, an interrupt index identifier
+is called interrupt number.
+
+**Interrupt management API functions**
+
+Interrupt management resources are declared in header file
+kernel/interrupt.h_. Interrupt consumers main API functions are:
+
+    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
+      enable and disable an interrupt.
```
Comment (jenswi-linaro) [inline architecture/core.rst:415]: enable or disable?

Comment (etienne-lms) [reply architecture/core.rst:415]: ok; s/and/or/

File: architecture/core.rst (hunk 4 above)
Comment (etienne-lms) [reply architecture/core.rst:418]: ok, i'll add a comment.

File: architecture/core.rst (hunk 3 above)
Comment (etienne-lms) [reply architecture/core.rst:454]: ok, thanks.

File: architecture/core.rst [hunk 6]
```diff
@@ -443,19 +444,19 @@ consumer(s) handler(s) with API function ``interrupt_call_handlers()``.
...
 reference with API function ``interrupt_get_main_chip()``.
 
 **Interrupt handlers**
 
-Interrupt handler functions are functions registered by interrupt consumer
-drivers that core shall call when the related interrupt occurs. Structure
-``struct itr_handler`` references a handler. It contains the handler function
-entry point, the interrupt number, the interrupt controller device and a
-few more parameters.
+Interrupt handler functions are are callback functions registered by interrupt
```
Comment (jenswi-linaro) [inline architecture/core.rst:455]: s/are are/are/

File: architecture/core.rst [hunk 7]
```diff
@@ -412,10 +412,11 @@ Interrupt management resources are declared in header file
 kernel/interrupt.h_. Interrupt consumers main API functions are:
 
     - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
-      enable and disable an interrupt.
+      enable or disable an interrupt.
 
     - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
-      mask/unmask an interrupt.
+      mask or unmask an interrupt. Masking an interrupt keeps it enabled
+      but disables the asynchronus notification of the event.
```
Comment (jenswi-linaro) [inline architecture/core.rst:419]: That's not quite true. They are actually identical, mask/unmask are implemented using disable/enable in all upstream drivers. How about something like:

Masking of an enabled interrupt temporarily disables the interrupt while unmasking enables a previously masked interrupt. `interrpt_mask()` and `interrpt_unmask()` are allowed to be called from an interrupt context, but `interrupt_enable()` and `interrupt_disable()` not so.

File: architecture/core.rst [hunk 8]
```diff
@@ -415,8 +415,11 @@ kernel/interrupt.h_. Interrupt consumers main API functions are:
       enable or disable an interrupt.
 
     - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
-      mask or unmask an interrupt. Masking an interrupt keeps it enabled
-      but disables the asynchronus notification of the event.
+      mask or unmask an interrupt.
+      Masking of an enabled interrupt temporarily disables the interrupt
+      while unmasking enables a previously masked interrupt. interrpt_mask()
```
Comment (jenswi-linaro) [inline architecture/core.rst:420]: It seems that you copied my typos. s/interrpt_/interrupt_/ on this line and on the line below

**Review by jenswi-linaro — COMMENTED** [review]
With my last comment addressed:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Review by etienne-lms — COMMENTED** [review]
I've sqaushed the fixup commits + the change below (missing quotes around function labels), and applied your review tag. Thanks for the review.

```patch
     - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
-      mask or unmask an interrupt.
-      Masking of an enabled interrupt temporarily disables the interrupt
-      while unmasking enables a previously masked interrupt. interrpt_mask()
-      and interrpt_unmask() are allowed to be called from an interrupt
-      context, but interrupt_enable() and interrupt_disable() not so.
+      mask or unmask an interrupt. Masking of an enabled interrupt
+      temporarily disables the interrupt while unmasking enables a
+      previously masked interrupt. ``interrupt_mask()`` and
+      ``interrupt_unmask()`` are allowed to be called from an interrupt
+      context, but ``interrupt_enable()`` and ``interrupt_disable()`` not so.
 ```
Note: the last 'so' sound strange to me.

**Comment by jbech-linaro** [comment]
Not sure why this pass on readthedocs, since I see both Error and warnings when looking that the readthedocs log. I noticed that when building this locally:
```
/home/jyx/devel/optee_projects/optee_docs/architecture/core.rst:419: ERROR: Unexpected indentation.
/home/jyx/devel/optee_projects/optee_docs/architecture/core.rst:411: ERROR: Unknown target name: "interrupt.h".
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index
/home/jyx/devel/optee_projects/optee_docs/architecture/core.rst:468: WARNING: undefined label: '_notifications'
```

**Comment by etienne-lms** [comment]
Thanks @jbech-linaro. Indeed I forgot to test building the doc.
Now it should be ok with the appended fixup commit.

File: architecture/core.rst [hunk 9]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
 several trusted threads in optee_os helps normal world scheduler to be
 efficient.
 
+Core handlers for native interrupts
+===================================
+
+OP-TEE core provides methods for device drivers to setup and register
+handler functions for native interrupt controller drivers
+(see:ref:`native_foreign_irqs`).
+Interrupt handlers can be nested for when an interrupt controller
```
Comment (jbech-linaro) [inline architecture/core.rst:399]: The `for` probably shouldn't be there, what about changing 399-401  to:
_Interrupt handlers can be nested when an interrupt controller exposes interrupts, which are multiplexed on an interrupt controlled by a parent interrupt controller._

File: architecture/core.rst [hunk 10]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+
+    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
+      enable or disable an interrupt.
+
+    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively mask
+or unmask an interrupt. Masking of an enabled interrupt temporarily disables
+the interrupt while unmasking enables a previously masked interrupt.
+``interrupt_mask()`` and ``interrupt_unmask()`` are allowed to be called
+from an interrupt context, but ``interrupt_enable()`` and
+``interrupt_disable()`` not so.
```
Comment (jbech-linaro) [inline architecture/core.rst:422]: You indent all the other items in this list except this one. Please indent so they follow the same formatting as the rest.

File: architecture/core.rst [hunk 11]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+
+    - ``interrupt_configure()`` to configure an interrupt detection mode
+      and priority.
+
+    - ``interrupt_add_handler()`` to configure and interrupt and register
```
Comment (jbech-linaro) [inline architecture/core.rst:427]: s/to configure and interrupt/to configure **an** interrupt/

File: architecture/core.rst [hunk 12]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+      a interrupt handler function, see below.
```
Comment (jbech-linaro) [inline architecture/core.rst:428]: s/a interrupt/**an** interrupt/

File: architecture/core.rst [hunk 13]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+    - ``interrupt_remove_handler()`` to unregister an interrupt handler
+      function from an interrupt.
+
+**Interrupt controller drivers**
+
+An interrupt controller instance, named chip (``struct itr_chip``) defines
+operation function handlers for management of the interrupt(s) it controls.
+An interrupt chip driver must provide operation handler functions ``.add``,
+``.mask``, ``.unmask``, ``.enable`` and ``.disable``. There are other
+operation handler functions that are optional, as ``.rasie_pi``,
```
Comment (jbech-linaro) [inline architecture/core.rst:439]: s/as/as **for example**/

File: architecture/core.rst [hunk 14]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+``.raise_sgi`` and ``.set_priority``.
+
+An interrupt chip driver registers the controller instance with API
+function ``itr_chip_init()``. The driver calls the registered interrupt
+consumer(s) handler(s) with API function ``interrupt_call_handlers()``.
+
+**CPU main interrupt controller driver**
+
+CPU main interrupt controller (e.g. a GIC instance on Arm architecture CPUs)
```
Comment (jbech-linaro) [inline architecture/core.rst:448]: I leave this open, but is it
"CPU main interrupt controller"
or 
"The main CPU interrupt controller"
?

File: architecture/core.rst [hunk 15]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+driver must register itself as main controller using API function
+``interrupt_main_init()``. The function is in charge of calling
+``itr_chip_init()`` for that chip instance.
+
+Interrupt consumer drivers get can get the CPU main interrupt controller
+reference with API function ``interrupt_get_main_chip()``.
```
Comment (jbech-linaro) [inline architecture/core.rst:454]: One error (one 'get' too much).
s/... drivers get can get the ... / ... drivers can get the ... /

However, just a suggestion for a slight re-write, leave it up to you to decide.
s/
Interrupt consumer drivers get can get the CPU main interrupt controller reference with API function ...
/
Interrupt consumer drivers can get a reference to the CPU main interrupt controller with the API function ...
/

File: architecture/core.rst [hunk 16]
```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
...
+An interrupt handler function return value is of type ``enum itr_return``.
+It shall return ``ITRR_HANDLED`` when the interrupt is served and
+``ITRR_NONE`` when the interrupt cannot be served.
+
+The interrupt handler runs in an interrupt context rather than a thread
+context. When this occurs, all other interrupts are masked, necessitating fast
+execution of the interrupt handler to avoid delaying or missing out on other
+interrupts. If there are long lasting operations to do on that interrupt
+occurrence, the interrupt handler should request execution of the OP-TEE
+bottom half thread, see :ref:`notifications`.
```
Comment (jbech-linaro) [inline architecture/core.rst:473]: Yet another suggestion:

s/If there ... bottom half thread/When an interrupt occurs that requires the completion of long-running operations, the interrupt handler should request that the OP-TEE bottom half thread be executed./

Comment (etienne-lms) [reply architecture/core.rst:473]: Ok, I'll reword the end:

_When an interrupt occurs that requires the completion of long-running operations, the interrupt handler should request ~~that~~ the OP-TEE bottom half thread **to execute those operations**, see ..._

**Comment by etienne-lms** [comment]
comments addressed

**Comment by jbech-linaro** [comment]
Thanks @etienne-lms 

`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by etienne-lms** [comment]
@jenswi-linaro, I guess your review tag still applies.

**Comment by jenswi-linaro** [comment]
> @jenswi-linaro, I guess your review tag still applies.

Yes :-)

**Comment by etienne-lms** [comment]
fixup commits squashed
review tags applied
thanks for the fixes

**Pull request merged**